flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java

```diff
@@ -35,109 +46,111 @@
 	final Map<String, TaskManagerMetricStore> taskManagers = new HashMap<>();
 	final Map<String, JobMetricStore> jobs = new HashMap<>();
 
-	/**
-	 * Adds a metric to this MetricStore.
-	 *
-	 * @param name  the metric identifier
-	 * @param value the metric value
-	 */
-	public void add(String name, Object value) {
-		TaskManagerMetricStore tm;
-		JobMetricStore job;
-		TaskMetricStore task;
-
+	public void add(MetricDump metric) {
 		try {
-			String[] components = name.split(":");
-			switch (components[0]) {
-				/**
-				 * JobManagerMetricStore metric
-				 * format: 0:<user_scope>.<name>
-				 */
-				case "0":
-					jobManager.metrics.put(components[1], value);
-					break;
-				/**
-				 * TaskManager metric
-				 * format: 1:<tm_id>:<user_scope>.<name>
-				 */
-				case "1":
-					if (components.length != 3) {
-						break;
-					}
-					tm = taskManagers.get(components[1]);
+			QueryScopeInfo info = metric.scopeInfo;
+			TaskManagerMetricStore tm;
+			JobMetricStore job;
+			TaskMetricStore task;
+
+			String name = info.scope.isEmpty()
+				? metric.name
+				: info.scope + "." + metric.name;
+			
+			if (name.isEmpty()) { // malformed transmission
+				return;
+			}
+
+			switch (info.getCategory()) {
+				case INFO_CATEGORY_JM:
```

Comment:
eh, seemed like the proper way of handling it. Also, (up to) 4 comparisons vs a jump.